Dev loop fails before a build

Version: 2.0.0-M3
Steps to reproduce

```
>> helidon init --version 2.0.0-M2 --flavor mp
Creating helidon-quickstart-mp using version 2.0.0-M2
Switch directory to ./helidon-quickstart-mp to use CLI

>> cd ...; helidon dev

| building (full)
| build completed (1.8 seconds)
| helidon-quickstart-mp starting

Exception in thread "main" java.lang.NoClassDefFoundError: io/helidon/microprofile/server/Server
        at io.helidon.examples.quickstart.mp.Main.startServer(Main.java:59)
        at io.helidon.examples.quickstart.mp.Main.main(Main.java:46)
Caused by: java.lang.ClassNotFoundException: io.helidon.microprofile.server.Server
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:602)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        ... 2 more

```

--- a/dev-loop/src/main/java/io/helidon/build/dev/Project.java
+++ b/dev-loop/src/main/java/io/helidon/build/dev/Project.java
@@ -464,6 +464,7 @@ public class Project {
         dependencyPaths.forEach(this::addDependency);
 
         // Build/rebuild classPath, weeding out any duplicates
+        // First, add each java build root
 
         final Set<Path> paths = new LinkedHashSet<>();
         components.forEach(component -> {
@@ -471,7 +472,12 @@ public class Project {
                 paths.add(component.outputRoot().path());
             }
         });
+
+        // Add all dependencies
+
         dependencies.forEach(dependency -> paths.add(dependency.path()));
+
+        // Finally, set the classpath from all the paths we collected
 
         classPath.clear();
         paths.forEach(path -> classPath.add(path.toFile()));

--- a/dev-loop/src/main/java/io/helidon/build/dev/maven/MavenProjectConfigCollector.java
+++ b/dev-loop/src/main/java/io/helidon/build/dev/maven/MavenProjectConfigCollector.java
@@ -17,14 +17,12 @@
 package io.helidon.build.dev.maven;
 
 import java.nio.file.Path;
-import java.util.ArrayList;
 import java.util.List;
 import java.util.stream.Collectors;
 
 import io.helidon.build.util.ProjectConfig;
 
 import org.apache.maven.AbstractMavenLifecycleParticipant;
-import org.apache.maven.artifact.DependencyResolutionRequiredException;
 import org.apache.maven.execution.AbstractExecutionListener;
 import org.apache.maven.execution.ExecutionEvent;
 import org.apache.maven.execution.ExecutionListener;
@@ -38,7 +36,6 @@ import org.codehaus.plexus.component.annotations.Component;
 
 import static io.helidon.build.util.ProjectConfig.DOT_HELIDON;
 import static io.helidon.build.util.ProjectConfig.PROJECT_CLASSDIRS;
-import static io.helidon.build.util.ProjectConfig.PROJECT_CLASSPATH;
 import static io.helidon.build.util.ProjectConfig.PROJECT_MAINCLASS;
 import static io.helidon.build.util.ProjectConfig.PROJECT_RESOURCEDIRS;
 import static io.helidon.build.util.ProjectConfig.PROJECT_SOURCEDIRS;
@@ -56,14 +53,13 @@ public class MavenProjectConfigCollector extends AbstractMavenLifecycleParticipa
     private static final String MULTI_MODULE_PROJECT = "Multi-module projects are not supported.";
     private static final String MISSING_MAIN_CLASS = "The required '" + MAIN_CLASS_PROPERTY + "' property is missing.";
     private static final String MISSING_DOT_HELIDON = "The required " + DOT_HELIDON + " file is missing.";
-    private static final String COPY_DEPENDENCIES_GOAL = "copy-dependencies";
-    private static final String COPY_LIBS_EXECUTION_ID = "copy-libs";
-    private static final String TARGET_DIR_NAME = "target";
-    private static final String LIBS_DIR_NAME = "libs";
+    private static final List<String> COMPILE_AND_RUNTIME_SCOPES = List.of("compile", "runtime");
+    private static final List<String> COMPILE_SCOPE = List.of("compile");
+    private static final String COMPILE_GOAL = "compile";
     private static final String JAR_FILE_SUFFIX = ".jar";
 
     private Path supportedProjectDir;
-    private ProjectConfig config;
+    private ProjectConfig projectConfig;
 
     /**
      * Assert that the project is one whose configuration we can support.
@@ -85,7 +81,7 @@ public class MavenProjectConfigCollector extends AbstractMavenLifecycleParticipa
     public void afterProjectsRead(MavenSession session) {
         // Init state
         supportedProjectDir = null;
-        config = null;
+        projectConfig = null;
         try {
             // Ensure that we support this project
             supportedProjectDir = assertSupportedProject(session);
@@ -108,7 +104,7 @@ public class MavenProjectConfigCollector extends AbstractMavenLifecycleParticipa
             } else if (result.hasExceptions()) {
                 debug("Build failed: %s", result.getExceptions());
                 invalidateConfig();
-            } else if (config != null) {
+            } else if (projectConfig != null) {
                 debug("Build succeeded, with compilation. Updating config.");
                 storeConfig();
             } else {
@@ -119,40 +115,38 @@ public class MavenProjectConfigCollector extends AbstractMavenLifecycleParticipa
     }
 
     private void collectConfig(MavenProject project) {
-        try {
-            final Path projectDir = project.getBasedir().toPath();
-            config = ProjectConfig.loadHelidonCliConfig(projectDir);
-            config.property(PROJECT_MAINCLASS, project.getProperties().getProperty(MAIN_CLASS_PROPERTY));
-            config.property(PROJECT_VERSION, project.getVersion());
-            final List<String> classesDirs = project.getCompileClasspathElements()
-                                                    .stream()
-                                                    .filter(d -> !d.endsWith(JAR_FILE_SUFFIX))
-                                                    .collect(Collectors.toList());
-            config.property(PROJECT_CLASSDIRS, classesDirs);
-            config.property(PROJECT_SOURCEDIRS, project.getCompileSourceRoots());
-            final List<String> resourceDirs = project.getResources()
-                                                     .stream()
-                                                     .map(Resource::getDirectory)
-                                                     .collect(Collectors.toList());
-            config.property(PROJECT_RESOURCEDIRS, resourceDirs);
-            final Path libsDir = projectDir.resolve(TARGET_DIR_NAME).resolve(LIBS_DIR_NAME);
-            final List<String> classPath = new ArrayList<>(classesDirs);
-            classPath.add(libsDir.toString());
-            config.property(PROJECT_CLASSPATH, classPath);
-        } catch (DependencyResolutionRequiredException e) {
-            throw new RuntimeException(e);
-        }
+        final Path projectDir = project.getBasedir().toPath();
+        final ProjectConfig config = ProjectConfig.loadHelidonCliConfig(projectDir);
+        config.property(PROJECT_MAINCLASS, project.getProperties().getProperty(MAIN_CLASS_PROPERTY));
+        config.property(PROJECT_VERSION, project.getVersion());
+        final Path outputDir = projectDir.resolve(project.getBuild().getOutputDirectory());
+        final List<String> classesDirs = List.of(outputDir.toString());
+        config.property(PROJECT_CLASSDIRS, classesDirs);
+        config.property(PROJECT_SOURCEDIRS, project.getCompileSourceRoots());
+        final List<String> resourceDirs = project.getResources()
+                                                 .stream()
+                                                 .map(Resource::getDirectory)
+                                                 .collect(Collectors.toList());
+        config.property(PROJECT_RESOURCEDIRS, resourceDirs);
+
+        // NOTE: The classpath is computed by MavenProjectSupplier rather than being
+        //       computed and stored here. This was done to solve a bug with MP where
+        //       the classpath was incorrect at the end of the compile step, and the
+        //       fix to correct it here required the use of "prepare-package" rather
+        //       than "process-classes", slowing everything down.
+
+        this.projectConfig = config;
     }
 
     private void storeConfig() {
-        config.buildSucceeded();
-        config.store();
+        projectConfig.buildSucceeded();
+        projectConfig.store();
     }
 
     private void invalidateConfig() {
-        this.config = ProjectConfig.loadHelidonCliConfig(supportedProjectDir);
-        config.buildFailed();
-        config.store();
+        projectConfig = ProjectConfig.loadHelidonCliConfig(supportedProjectDir);
+        projectConfig.buildFailed();
+        projectConfig.store();
     }
 
     private static void assertSupportedProject(boolean supported, String reasonIfUnsupported) {
@@ -216,8 +210,7 @@ public class MavenProjectConfigCollector extends AbstractMavenLifecycleParticipa
         public void mojoSucceeded(ExecutionEvent event) {
             final MojoExecution execution = event.getMojoExecution();
             next.mojoSucceeded(event);
-            if (execution.getGoal().equals(COPY_DEPENDENCIES_GOAL)
-                && execution.getExecutionId().equals(COPY_LIBS_EXECUTION_ID)) {
+            if (execution.getGoal().equals(COMPILE_GOAL)) {
                 collectConfig(event.getProject());
             }
         }

--- a/dev-loop/src/main/java/io/helidon/build/dev/maven/MavenProjectConfigCollector.java
+++ b/dev-loop/src/main/java/io/helidon/build/dev/maven/MavenProjectConfigCollector.java
@@ -17,6 +17,7 @@
 package io.helidon.build.dev.maven;
 
 import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.stream.Collectors;
 
@@ -55,88 +56,101 @@ public class MavenProjectConfigCollector extends AbstractMavenLifecycleParticipa
     private static final String MULTI_MODULE_PROJECT = "Multi-module projects are not supported.";
     private static final String MISSING_MAIN_CLASS = "The required '" + MAIN_CLASS_PROPERTY + "' property is missing.";
     private static final String MISSING_DOT_HELIDON = "The required " + DOT_HELIDON + " file is missing.";
-    private static final String COMPILE_GOAL = "compile";
+    private static final String COPY_DEPENDENCIES_GOAL = "copy-dependencies";
+    private static final String COPY_LIBS_EXECUTION_ID = "copy-libs";
+    private static final String TARGET_DIR_NAME = "target";
+    private static final String LIBS_DIR_NAME = "libs";
+    private static final String JAR_FILE_SUFFIX = ".jar";
 
-    private boolean supportedProject;
-    private boolean compileSucceeded;
+    private Path supportedProjectDir;
+    private ProjectConfig config;
 
     /**
      * Assert that the project is one whose configuration we can support.
      *
      * @param session The session.
+     * @return The project directory.
      */
-    public static void assertSupportedProject(MavenSession session) {
+    public static Path assertSupportedProject(MavenSession session) {
         assertSupportedProject(session.getProjects().size() == 1, MULTI_MODULE_PROJECT);
         final MavenProject project = session.getProjects().get(0);
         final Path projectDir = project.getBasedir().toPath();
         assertSupportedProject(ProjectConfig.helidonCliConfigExists(projectDir), MISSING_DOT_HELIDON);
         assertSupportedProject(project.getProperties().getProperty(MAIN_CLASS_PROPERTY) != null, MISSING_MAIN_CLASS);
         debug("Helidon project is supported");
+        return projectDir;
     }
 
     @Override
     public void afterProjectsRead(MavenSession session) {
+        // Init state
+        supportedProjectDir = null;
+        config = null;
         try {
             // Ensure that we support this project
-            assertSupportedProject(session);
-            supportedProject = true;
+            supportedProjectDir = assertSupportedProject(session);
 
             // Install our listener so we can know if compilation occurred and succeeded
             final MavenExecutionRequest request = session.getRequest();
             request.setExecutionListener(new EventListener(request.getExecutionListener()));
         } catch (IllegalStateException e) {
-            supportedProject = false;
+            supportedProjectDir = null;
         }
     }
 
     @Override
     public void afterSessionEnd(MavenSession session) {
-        if (supportedProject) {
-            final MavenProject project = session.getProjects().get(0);
-            final Path projectDir = project.getBasedir().toPath();
-            final ProjectConfig config = ProjectConfig.loadHelidonCliConfig(projectDir);
+        if (supportedProjectDir != null) {
             final MavenExecutionResult result = session.getResult();
             if (result == null) {
                 debug("Build failed: no result");
-                invalidateConfig(config);
+                invalidateConfig();
             } else if (result.hasExceptions()) {
                 debug("Build failed: %s", result.getExceptions());
-                invalidateConfig(config);
-            } else if (compileSucceeded) {
+                invalidateConfig();
+            } else if (config != null) {
                 debug("Build succeeded, with compilation. Updating config.");
-                updateConfig(config, project);
+                storeConfig();
             } else {
                 debug("Build succeeded, without compilation");
-                invalidateConfig(config);
+                invalidateConfig();
             }
         }
     }
 
-    private void updateConfig(ProjectConfig config, MavenProject project) {
+    private void collectConfig(MavenProject project) {
         try {
-            final String mainClass = project.getProperties().getProperty(MAIN_CLASS_PROPERTY);
-            config.property(PROJECT_MAINCLASS, mainClass);
+            final Path projectDir = project.getBasedir().toPath();
+            config = ProjectConfig.loadHelidonCliConfig(projectDir);
+            config.property(PROJECT_MAINCLASS, project.getProperties().getProperty(MAIN_CLASS_PROPERTY));
             config.property(PROJECT_VERSION, project.getVersion());
-            config.property(PROJECT_CLASSPATH, project.getRuntimeClasspathElements());
-            config.property(PROJECT_SOURCEDIRS, project.getCompileSourceRoots());
             final List<String> classesDirs = project.getCompileClasspathElements()
                                                     .stream()
-                                                    .filter(d -> !d.endsWith(".jar"))
+                                                    .filter(d -> !d.endsWith(JAR_FILE_SUFFIX))
                                                     .collect(Collectors.toList());
             config.property(PROJECT_CLASSDIRS, classesDirs);
+            config.property(PROJECT_SOURCEDIRS, project.getCompileSourceRoots());
             final List<String> resourceDirs = project.getResources()
                                                      .stream()
                                                      .map(Resource::getDirectory)
                                                      .collect(Collectors.toList());
             config.property(PROJECT_RESOURCEDIRS, resourceDirs);
-            config.buildSucceeded();
-            config.store();
+            final Path libsDir = projectDir.resolve(TARGET_DIR_NAME).resolve(LIBS_DIR_NAME);
+            final List<String> classPath = new ArrayList<>(classesDirs);
+            classPath.add(libsDir.toString());
+            config.property(PROJECT_CLASSPATH, classPath);
         } catch (DependencyResolutionRequiredException e) {
             throw new RuntimeException(e);
         }
     }
 
-    private void invalidateConfig(ProjectConfig config) {
+    private void storeConfig() {
+        config.buildSucceeded();
+        config.store();
+    }
+
+    private void invalidateConfig() {
+        this.config = ProjectConfig.loadHelidonCliConfig(supportedProjectDir);
         config.buildFailed();
         config.store();
     }
@@ -202,8 +216,9 @@ public class MavenProjectConfigCollector extends AbstractMavenLifecycleParticipa
         public void mojoSucceeded(ExecutionEvent event) {
             final MojoExecution execution = event.getMojoExecution();
             next.mojoSucceeded(event);
-            if (execution.getGoal().equals(COMPILE_GOAL)) {
-                compileSucceeded = true;
+            if (execution.getGoal().equals(COPY_DEPENDENCIES_GOAL)
+                && execution.getExecutionId().equals(COPY_LIBS_EXECUTION_ID)) {
+                collectConfig(event.getProject());
             }
         }
 

--- a/dev-loop/src/main/java/io/helidon/build/dev/maven/MavenProjectSupplier.java
+++ b/dev-loop/src/main/java/io/helidon/build/dev/maven/MavenProjectSupplier.java
@@ -24,7 +24,6 @@ import java.nio.file.Path;
 import java.nio.file.attribute.BasicFileAttributes;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicBoolean;
-import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 
 import io.helidon.build.dev.BuildExecutor;
@@ -35,6 +34,7 @@ import io.helidon.build.dev.BuildType;
 import io.helidon.build.dev.DirectoryType;
 import io.helidon.build.dev.FileType;
 import io.helidon.build.dev.Project;
+import io.helidon.build.dev.Project.Builder;
 import io.helidon.build.dev.ProjectDirectory;
 import io.helidon.build.dev.ProjectSupplier;
 import io.helidon.build.util.Log;
@@ -42,7 +42,11 @@ import io.helidon.build.util.ProjectConfig;
 
 import org.apache.maven.execution.MavenSession;
 import org.apache.maven.plugin.BuildPluginManager;
+import org.apache.maven.project.DefaultDependencyResolutionRequest;
+import org.apache.maven.project.DependencyResolutionException;
+import org.apache.maven.project.DependencyResolutionRequest;
 import org.apache.maven.project.MavenProject;
+import org.apache.maven.project.ProjectDependenciesResolver;
 
 import static io.helidon.build.dev.BuildComponent.createBuildComponent;
 import static io.helidon.build.dev.BuildFile.createBuildFile;
@@ -54,19 +58,22 @@ import static io.helidon.build.util.FileUtils.assertFile;
 import static io.helidon.build.util.FileUtils.ensureDirectory;
 import static io.helidon.build.util.FileUtils.lastModifiedTime;
 import static io.helidon.build.util.ProjectConfig.PROJECT_CLASSDIRS;
-import static io.helidon.build.util.ProjectConfig.PROJECT_CLASSPATH;
 import static io.helidon.build.util.ProjectConfig.PROJECT_MAINCLASS;
 import static io.helidon.build.util.ProjectConfig.PROJECT_RESOURCEDIRS;
 import static io.helidon.build.util.ProjectConfig.PROJECT_SOURCEDIRS;
 import static io.helidon.build.util.ProjectConfig.loadHelidonCliConfig;
 import static io.helidon.build.util.TimeUtils.toDateTime;
+import static java.util.Objects.requireNonNull;
+import static org.eclipse.aether.util.artifact.JavaScopes.COMPILE;
+import static org.eclipse.aether.util.artifact.JavaScopes.RUNTIME;
+import static org.eclipse.aether.util.filter.DependencyFilterUtils.classpathFilter;
 
 /**
  * A {@code ProjectSupplier} for Maven projects.
  */
 public class MavenProjectSupplier implements ProjectSupplier {
-    private static final List<String> CLEAN_BUILD_COMMAND = List.of("clean", "prepare-package", "-DskipTests");
-    private static final List<String> BUILD_COMMAND = List.of("prepare-package", "-DskipTests");
+    private static final List<String> CLEAN_BUILD_COMMAND = List.of("clean", "process-classes", "-DskipTests");
+    private static final List<String> BUILD_COMMAND = List.of("process-classes", "-DskipTests");
     private static final String TARGET_DIR_NAME = "target";
     private static final String POM_FILE = "pom.xml";
     private static final String DOT = ".";
@@ -81,9 +88,11 @@ public class MavenProjectSupplier implements ProjectSupplier {
         return !name.equals(TARGET_DIR_NAME);
     };
 
-    private final AtomicReference<MavenProject> project;
+    private final MavenProject project;
     private final MavenSession session;
     private final BuildPluginManager plugins;
+    private final ProjectDependenciesResolver dependencyResolver;
+    private final DependencyResolutionRequest dependencyRequest;
     private final AtomicBoolean firstBuild;
     private BuildType buildType;
     private ProjectConfig config;
@@ -167,14 +176,19 @@ public class MavenProjectSupplier implements ProjectSupplier {
      * @param project The maven project.
      * @param session The maven session.
      * @param plugins The maven plugin manager.
+     * @param resolver The maven dependencies resolver;.
      */
     public MavenProjectSupplier(MavenProject project,
                                 MavenSession session,
-                                BuildPluginManager plugins) {
+                                BuildPluginManager plugins,
+                                ProjectDependenciesResolver resolver) {
         MavenProjectConfigCollector.assertSupportedProject(session);
-        this.project = new AtomicReference<>(project);
-        this.session = session;
-        this.plugins = plugins;
+        this.project = project;
+        this.session = requireNonNull(session);
+        this.plugins = requireNonNull(plugins);
+        this.dependencyResolver = requireNonNull(resolver);
+        this.dependencyRequest = new DefaultDependencyResolutionRequest(project, requireNonNull(session.getRepositorySession()))
+            .setResolutionFilter(classpathFilter(COMPILE, RUNTIME));
         this.firstBuild = new AtomicBoolean(true);
     }
 
@@ -231,7 +245,7 @@ public class MavenProjectSupplier implements ProjectSupplier {
 
     private Project createProject(Path projectDir, BuildType buildType) {
         // Root directory
-        final Project.Builder builder = Project.builder().buildType(buildType);
+        final Builder builder = Project.builder().buildType(buildType);
         final ProjectDirectory root = createProjectDirectory(DirectoryType.Project, projectDir);
         builder.rootDirectory(root);
 
@@ -239,9 +253,8 @@ public class MavenProjectSupplier implements ProjectSupplier {
         final Path pomFile = assertFile(projectDir.resolve(POM_FILE));
         builder.buildFile(createBuildFile(root, FileType.MavenPom, pomFile));
 
-        // Dependencies - Should we consider the classes/lib?
-        final List<String> classpath = config.propertyAsList(PROJECT_CLASSPATH);
-        classpath.stream().map(s -> Path.of(s)).forEach(builder::dependency);
+        // Dependencies
+        addDependencies(builder);
 
         // Build components
         final List<String> sourceDirs = config.propertyAsList(PROJECT_SOURCEDIRS);
@@ -276,9 +289,22 @@ public class MavenProjectSupplier implements ProjectSupplier {
         return builder.build();
     }
 
+    private void addDependencies(Builder builder) {
+        try {
+            dependencyResolver.resolve(dependencyRequest)
+                              .getDependencies()
+                              .stream()
+                              .map(d -> d.getArtifact().getFile().toPath())
+                              .forEach(builder::dependency);
+        } catch (DependencyResolutionException e) {
+            throw new IllegalStateException("Dependency resolution failed: " + e.getMessage());
+        }
+    }
+
+
     private BuildStep compileStep() {
         return MavenGoalBuildStep.builder()
-                                 .mavenProject(project.get())
+                                 .mavenProject(project)
                                  .mavenSession(session)
                                  .pluginManager(plugins)
                                  .goal(MavenGoalBuildStep.compileGoal())
@@ -287,7 +313,7 @@ public class MavenProjectSupplier implements ProjectSupplier {
 
     private BuildStep resourcesStep() {
         return MavenGoalBuildStep.builder()
-                                 .mavenProject(project.get())
+                                 .mavenProject(project)
                                  .mavenSession(session)
                                  .pluginManager(plugins)
                                  .goal(MavenGoalBuildStep.resourcesGoal())

--- a/dev-loop/src/main/java/io/helidon/build/dev/maven/MavenProjectSupplier.java
+++ b/dev-loop/src/main/java/io/helidon/build/dev/maven/MavenProjectSupplier.java
@@ -65,8 +65,8 @@ import static io.helidon.build.util.TimeUtils.toDateTime;
  * A {@code ProjectSupplier} for Maven projects.
  */
 public class MavenProjectSupplier implements ProjectSupplier {
-    private static final List<String> CLEAN_BUILD_COMMAND = List.of("clean", "process-classes", "-DskipTests");
-    private static final List<String> BUILD_COMMAND = List.of("process-classes", "-DskipTests");
+    private static final List<String> CLEAN_BUILD_COMMAND = List.of("clean", "prepare-package", "-DskipTests");
+    private static final List<String> BUILD_COMMAND = List.of("prepare-package", "-DskipTests");
     private static final String TARGET_DIR_NAME = "target";
     private static final String POM_FILE = "pom.xml";
     private static final String DOT = ".";

--- a/dev-loop/src/main/java/io/helidon/build/dev/mode/DevLoop.java
+++ b/dev-loop/src/main/java/io/helidon/build/dev/mode/DevLoop.java
@@ -145,7 +145,11 @@ public class DevLoop {
                 log("%s", BoldBlue.apply("up to date"));
             } else {
                 String operation = cycleNumber == 0 ? DEV_LOOP_BUILD_STARTING : "re" + DEV_LOOP_BUILD_STARTING;
-                log("%s (%s)", BoldBlue.apply(operation), type);
+                if (type == BuildType.Incremental) {
+                    log("%s (%s)", BoldBlue.apply(operation), type);
+                } else {
+                    log("%s", BoldBlue.apply(operation));
+                }
                 buildStartTime = System.currentTimeMillis();
             }
         }

--- a/helidon-maven-plugin/src/main/java/io/helidon/build/maven/dev/DevMojo.java
+++ b/helidon-maven-plugin/src/main/java/io/helidon/build/maven/dev/DevMojo.java
@@ -35,6 +35,7 @@ import org.apache.maven.plugins.annotations.Mojo;
 import org.apache.maven.plugins.annotations.Parameter;
 import org.apache.maven.plugins.annotations.ResolutionScope;
 import org.apache.maven.project.MavenProject;
+import org.apache.maven.project.ProjectDependenciesResolver;
 
 /**
  * Maven plugin that runs a {@link DevLoop}.
@@ -92,6 +93,12 @@ public class DevMojo extends AbstractMojo {
     @Component
     private BuildPluginManager plugins;
 
+    /**
+     * The Maven ProjectDependenciesResolver component.
+     */
+    @Component
+    private ProjectDependenciesResolver resolver;
+
     @Override
     public void execute() throws MojoExecutionException {
         if (skip) {
@@ -104,7 +111,7 @@ public class DevMojo extends AbstractMojo {
             } else {
                 MavenLogWriter.bind(getLog());
             }
-            final ProjectSupplier projectSupplier = new MavenProjectSupplier(project, session, plugins);
+            final ProjectSupplier projectSupplier = new MavenProjectSupplier(project, session, plugins, resolver);
             final DevLoop loop = new DevLoop(devProjectDir.toPath(), projectSupplier, clean, fork, terminalMode);
             loop.start(Integer.MAX_VALUE);
         } catch (Exception e) {

--- a/utils/src/main/java/io/helidon/build/util/ConfigProperties.java
+++ b/utils/src/main/java/io/helidon/build/util/ConfigProperties.java
@@ -22,6 +22,7 @@ import java.io.FileWriter;
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.util.Arrays;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
@@ -99,12 +100,12 @@ public class ConfigProperties {
     }
 
     /**
-     * Set property using string list and {@link #DELIMITER}.
+     * Set property using string collection and {@link #DELIMITER}.
      *
      * @param key The key.
-     * @param values The string list.
+     * @param values The strings.
      */
-    public void property(String key, List<String> values) {
+    public void property(String key, Collection<String> values) {
         properties.setProperty(key, String.join(DELIMITER, values));
     }
 

--- a/utils/src/main/java/io/helidon/build/util/ConfigProperties.java
+++ b/utils/src/main/java/io/helidon/build/util/ConfigProperties.java
@@ -170,7 +170,7 @@ public class ConfigProperties {
     public void store() {
         try {
             try (FileWriter writer = new FileWriter(file)) {
-                properties.store(writer, getClass().getName());
+                properties.store(writer, "Helidon Project Configuration");
             }
         } catch (IOException e) {
             throw new UncheckedIOException(e);

